eclipse-theia/theia
# Refactor communication layer to use socket.io

#### What it does

Closes #10403

Replaces most of our communication layer (previously using simple websockets + http-fallback) with the [socket.io](https://socket.io/) framework.

This removes the need for a custom long polling implementation, which is done by socket.io automatically. It works by first initiating a long-polling connection and after a successful handshake tries to create a websocket connection. If that fails, it will simply continue using the long-polling connection. This greatly increases the performance of the http-fallback, which now seems to be on par with the websocket solution (performance wise). For more info on the actual protocol used here, see [`socket.io-protocol`](https://github.com/socketio/socket.io-protocol) repo.

There is some weird behavior in socket.io that strips `origin` headers for some reason. This is undocumented and seems to be used by the cors feature. I've worked around this for now using a `fix-origin` header that gets transmitted to the backend and can then be used for validation. Hopefully the comments around these parts explain it well enough.

I've tried to keep the APIs as backwards compatible as possible. Hopefully this change shouldn't break too much. 

#### How to test

1. Start the browser application and assert that the frontend connects and everything works as expected
2. Do the same for electron
3. Also execute the testing steps for the http-fallback (#9731)

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
